docs/sat-solver.md

```diff
@@ -0,0 +1,383 @@
+<img src="https://render.githubusercontent.com/render/math?math=P1_{1}">
+version and the newest version
+<img src="https://render.githubusercontent.com/render/math?math=P1_{newest}">
+available in the repositories. This distance is a
+constant for the purposes of the PBO problem. In our case, we can define it as:
+
+<img src="https://render.githubusercontent.com/render/math?math=d(P1_{1}) = P1_{newest} * \overline{semver_{newest}} - P1_{1} * \overline{semver_{1}}">
+
+Where 
+<img src="https://render.githubusercontent.com/render/math?math=\overline{semver_{k}}">
+with version _k_ in the form of Helm's [semantic version
+ranges](https://devhints.io/semver).
+
+The normalized semantic version is calculated by constructing a fractional
+number such that:
+
+- `major version -> integer part`
+- `minor.patch version -> fractional part`
+
+E.g:
+
+- `1.2.1 -> 1,21`
+- `0.46.3 -> 0,463`
+- `7.10.1 -> 7,101`
+
+
+The objective function will be:
+
+<img src="https://render.githubusercontent.com/render/math?math=f_{3} = min( (P1_{1} * d(P1_{1}) %2B \dots %2B  P1_{k1} * d(P1_{k1}) ) %2B  (PZ_{1} * d(PZ_{1}) %2B \dots %2B  PZ_{kn} * d(PZ_{kn}) ))">
+
+where the value of 
+<img src="https://render.githubusercontent.com/render/math?math=d(Pi_{ki})">
+is zero if the package is the newest in the repo.
+
+
+### Multicriteria optimization
+
+Usually, installing a package follows multiple of the already mentioned
+criteria. Even if one is more important than the others, not following several
+of them can lead to undesired solutions (e.g: minimizing the amount of
+installed packages, yet wanting all optional dependencies to be installed).
+
+Given that each criteria has a different utility for the user, we can specify a
+weight
+<img src="https://render.githubusercontent.com/render/math?math=W_{i})">
+for each criteria _i_ (e.g: weight given to removals of packages, weight given
+to the present of packages, weight given to having an older version in the
+solution when a newer exists).
+
+Then, the objective function is defined as:
+
+<img src="https://render.githubusercontent.com/render/math?math=f_{4} = min( (W_{1} * f_{1}(P) %2B (W_{2} * f_{2}(P)) %2B (W_{3} * f_{3}(P))">
+
+# Existing solver libraries in Golang that we considered
+
+Ideally, we want a MAXSAT/Pseudo-Boolean SAT solver library with ample feature
+support, good docs, good API, a thriving community, verbose output, in Golang,
+and with a compatible license that can be statically linked. We want a solver
+that is used in the field, updated continuously with latest SAT improvements
+from SAT competitions. We want a solver that handles optimization problems
+(hence, a Pseudo-Boolean SAT solver).
+
+## [Gini](https://github.com/IRIFrance/gini)
+
+- 1 big known consumer: github.com/operator-framework/operator-lifecycle-manager.
+- Only pure SAT CNF problems.
+
+## [Luet](https://github.com/mudler/luet)
+
+- GPLv3, which would force us to move to GPLv3 given that Golang is usually
```

Comment:
Since we know the author of luet, I'd not consider the license issue a major one. :wink:

Reply:
True, we can reword that :). But notice that Luet is just codifying the problem in clauses, which are fed into gophersat. We will need to do that anyways, with more constraints than them (optional dependencies, semvers, provides, etc).

Reply:
reworded in https://github.com/rancher-sandbox/hypper/commit/54ac8ea6ccbc07c0181ad76a46812f70767bab4d.